### PR TITLE
Refactored UHFQuantumController

### PR DESCRIPTION
**Pull request by nielshaandbaek**
The purpose of this PR is to refactor the `UHFQuantumController` module in a similar fashion as was done for the HDAWG. As a result, a new module has been created named `UHFQA_core` containing a class of the same name with the core functionality of the driver. The `UHFQC`  class has been simplified accordingly and is now derived from `UHFQA_core`. Additionally, a `UHFQA_qudev` module has been created with Qudev specific functionality.

Fixes #194.

Changes proposed in this pull request:
- Created `UHFQA_core` module with core functionality.
- Simplified `UHFQuantumController` removing core functionality.
- Added Qudev specific functionality in `UHFQA_qudev`.

@MiguelSMoreira @chellings

In order for the pull request to be merged, the following conditions must be met:
- test suite (Github actions) passes
- all reasonable issues raised by codacy must be resolved 
- a positive review is required

Whenever possible the pull request should
- follow the PEP8 style guide 
- have tests for the code
- be well documented and contain comments

Tests are not mandatory as this is generally hard to make for instruments that interact with hardware.

### CODE REVIEW

**Review by chellings — CHANGES_REQUESTED** [review]
Thank you Niels for refactoring the UHF driver. I tested the new driver both with a mock server and on two different real setups, and it seems to work fine apart from the comments that I have added to this review.

Note that there will be merge conflicts with PR #641, which is approved, but not merged yet. Sorry, I am not sure about the policies in the Delft github: who will take care of merging after a pull request is approved? Once #641 got merged, I could push the conflict resolution to #656 since I have already resolved the conflicts locally.

File: pycqed/instrument_drivers/physical_instruments/ZurichInstruments/ZI_base_instrument.py
Comment (nielshaandbaek) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/ZI_base_instrument.py:1397]: Thanks, fixed.

File: pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py
Comment (nielshaandbaek) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py:61]: Thanks, fixed.

File: pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQuantumController.py
Comment (nielshaandbaek) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQuantumController.py:343]: Yes, hold-off errors are detected and reported by the hardware. Or they should be.

File: pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py
Comment (nielshaandbaek) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py:5]: That is a very good point, but I think we should address these issues in a separate ticket.

File: pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py
Comment (nielshaandbaek) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py:505]: Thanks, moved to `UHFQA_core`.

File: pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py
Comment (chellings) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py:5]: Thank you, Niels, for implementing the feedback. It looks good, except this aspect, where my comment was maybe not clear enough. The UHFQA_core that you have now merged is not functional because of this issue. The Delft-specific child class should work, but an object directly instantiated from the core (or from the qudev child class) won't work. An error will be raised when trying to run an experiment. I have used the following hack in our repository to make it work. Not sure whether there is a cleaner solution.

```
@@ -1102,22 +1102,24 @@ class ZI_base_instrument(Instrument):
             return np.genfromtxt(filename, delimiter=',')
         except OSError as e:
             # if the waveform does not exist yet dont raise exception
             log.warning(e)
             return None
 
     def _length_match_waveforms(self, awg_nr):
         """
         Adjust the length of a codeword waveform such that each individual
         waveform of the pair has the same length
         """
+        if not hasattr(self, '_get_waveform_table'):
+            return
         log.info('Length matching waveforms for dynamic waveform upload.')
         wf_table = self._get_waveform_table(awg_nr)
 
         matching_updated = False
         iter_id = 0
         # We iterate over the waveform table
         while(matching_updated or iter_id == 0):
 
             iter_id += 1
             if iter_id > 10:
                 raise StopIteration
@@ -1196,22 +1198,24 @@ class ZI_base_instrument(Instrument):
         if wf_name not in self._awg_waveforms:
             raise ziConfigurationError(
                 'Trying to mark waveform {} as read-only, but the waveform has not been configured yet!'.format(wf_name))
 
         self._awg_waveforms[wf_name]['readonly'] = True
 
     def _upload_updated_waveforms(self, awg_nr):
         """
         Loop through all configured waveforms and use dynamic waveform uploading
         to update changed waveforms on the instrument as needed.
         """
+        if not hasattr(self, '_get_waveform_table'):
+            return
         # Fixme. the _get_waveform_table should also be implemented for the UFH
         log.info(f"{self.devname}: Using dynamic waveform update for AWG {awg_nr}.")
         wf_table = self._get_waveform_table(awg_nr)
 
         for dio_cw, (wf_name, other_wf_name) in enumerate(wf_table):
             if self._awg_waveforms[wf_name]['dirty'] or self._awg_waveforms[other_wf_name]['dirty']:
                 # Combine the waveforms and upload
                 wf_data = merge_waveforms(self._awg_waveforms[wf_name]['waveform'],
                                           self._awg_waveforms[other_wf_name]['waveform'])
                 # Write the new waveform
                 self.setv(
```

Comment (nielshaandbaek) [reply pycqed/instrument_drivers/physical_instruments/ZurichInstruments/UHFQA_core.py:5]: See https://github.com/DiCarloLab-Delft/PycQED_py3/pull/657 for a possible solution.

**Pull request merged**
